dwst/scripts/test/particles.js

```diff
@@ -197,27 +197,154 @@ describe('particles module', () => {
-      const invalidParticlesExamples = [
-        '$',
-        '$ {foo()}',
-        '$foo()',
-        '${foo(123,)}',
-        '${foo(,456)}',
-        '${foo}',
-        '${foo(}',
-        '${foo(123',
-        '${foo(123,',
-        '${foo(123,456',
-        '${foo)}',
-        '${foo()',
-        '${foo}()}',
-        '\\a',
-      ];
-      invalidParticlesExamples.forEach(invalidExample => {
-        expect(() => {
-          return parseParticles(invalidExample);
-        }).to.throw(InvalidParticles);
+    it('should throw InvalidParticles for lone expression start', () => {
+      expect(() => {
+        return parseParticles('$');
+      }).to.throw(InvalidParticles).that.does.include({
+        expression: '$',
+        expected: '{',
+        remainder: '',
+        errorPosition: '$'.length,
+      });
+    });
+    it('should throw InvalidParticles for space between expression start and expression open', () => {
+      expect(() => {
+        return parseParticles('$ {foo()}');
+      }).to.throw(InvalidParticles).that.does.include({
+        expression: '$ {foo()}',
+        expected: '{',
+        remainder: ' {foo()}',
+        errorPosition: '$'.length,
+      });
+    });
+    it('should throw InvalidParticles for missing expression open and close', () => {
+      expect(() => {
+        return parseParticles('$foo()');
+      }).to.throw(InvalidParticles).that.does.include({
+        expression: '$foo()',
+        expected: '{',
+        remainder: 'foo()',
+        errorPosition: '$'.length,
+      });
+    });
+    it('should throw InvalidParticles for missing argument after a comma', () => {
+      expect(() => {
+        return parseParticles('${foo(123,)}');
+      }).to.throw(InvalidParticles).that.does.include({
+        expression: '${foo(123,)}',
+        expected: 'argument',
+        remainder: ')}',
+        errorPosition: '${foo(123,'.length,
+      });
+    });
+    it('should throw InvalidParticles for missing first argument before a comma', () => {
+      expect(() => {
+        return parseParticles('${foo(,456)}');
+      }).to.throw(InvalidParticles).that.does.include({
+        expression: '${foo(,456)}',
+        expected: 'argument',
+        remainder: ',456)}',
+        errorPosition: '${foo('.length,
+      });
+    });
+    it('should throw InvalidParticles for no argument list', () => {
+      expect(() => {
+        return parseParticles('${foo}');
+      }).to.throw(InvalidParticles).that.does.include({
+        expression: '${foo}',
+        expected: '(',
+        remainder: '}',
+        errorPosition: '${foo'.length,
+      });
+    });
+    it('should throw InvalidParticles for empty argument list missing close', () => {
+      expect(() => {
+        return parseParticles('${foo(}');
+      }).to.throw(InvalidParticles).that.does.include({
+        expression: '${foo(}',
+        expected: ', or )',
+        remainder: '}',
```

Comment:
This test fails with:
```
      AssertionError: expected [Error] to have property 'remainder' of '}', but got ''
      + expected - actual

      +}
```
This is why:
```
function readInstructionArg(parsee) {
  const arg = parsee.readUntil([' ', ',', ')']); // readUntil will read '}' into arg!
  if (arg.length === 0) {
    throw new InvalidParticles('argument', String(parsee));
  }
  if (parsee.length === 0) { // function will enter this block, '}' is missing form parsee, it was read to arg.
    throw new InvalidParticles(', or )', String(parsee));
  }
  return arg;
}
```
Maybe '}' should be a stopper?


Then we could do:
```
if (parsee.StartsWith('}') === false) {
  //throw error
}
```

Reply:
I think you need to define a white list for characters that are allowed in instruction arguments and use readWhile instead of readUntil.